api/lib/opentelemetry/trace/span_data.rb

```diff
@@ -0,0 +1,49 @@
+# frozen_string_literal: true
+
+# Copyright 2019 OpenTelemetry Authors
+#
+# SPDX-License-Identifier: Apache-2.0
+
+module OpenTelemetry
+  module Trace
+    # SpanData is an immutable object that is used to report out-of-band completed spans.
+    #
+    # TODO: consider whether to copy collections to a known internal form and expose only enumerations.
```

Comment:
I really like this idea. I think the no-op implementation should try to expose the smallest possible interface. That will help users ensure their usage is correct (e.g. they aren't inadvertently depending on the returned objects implementing obscure methods of Array or Hash), and relieve implementors of the burden of implementing a larger surface.

Reply:
Opened #15 for this. I don't want to invest effort until there's some clarity around `SpanData` (https://github.com/open-telemetry/opentelemetry-specification/issues/71) and enumerable vs. hash attributes (https://github.com/open-telemetry/opentelemetry-specification/issues/92).